roofit/roofitcore/inc/TestStatistics/RooSumL.h

```diff
@@ -0,0 +1,50 @@
+// Author: Patrick Bos, Netherlands eScience Center / NIKHEF 2021
+
+/*****************************************************************************
+ * RooFit
+ * Authors:                                                                  *
+ *   WV, Wouter Verkerke, UC Santa Barbara, verkerke@slac.stanford.edu       *
+ *   DK, David Kirkby,    UC Irvine,         dkirkby@uci.edu                 *
+ *                                                                           *
+ * Copyright (c) 2000-2021, Regents of the University of California          *
+ *                          and Stanford University. All rights reserved.    *
+ *                                                                           *
+ * Redistribution and use in source and binary forms,                        *
+ * with or without modification, are permitted according to the terms        *
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+
+#ifndef ROOT_ROOFIT_TESTSTATISTICS_RooSumL
+#define ROOT_ROOFIT_TESTSTATISTICS_RooSumL
+
+#include <TestStatistics/RooAbsL.h>
+#include <TestStatistics/optional_parameter_types.h>
+
+namespace RooFit {
+namespace TestStatistics {
+
+class RooSumL : public RooAbsL {
+public:
+   // main constructor
+   RooSumL(RooAbsPdf* pdf, RooAbsData* data, std::vector<std::unique_ptr<RooAbsL>> components,
+           RooAbsL::Extended extended = RooAbsL::Extended::Auto);
+   // Note: when above ctor is called without std::moving components, you get a really obscure error. Pass as std::move(components)!
```

Comment:
Try this in godbolt - it doesn't compile:
```c++
#include <vector>
#include <memory>
#include <cstdio>

void acceptUniquePtr(std::vector<std::unique_ptr<int>> vec) {
    printf("number is %d", *vec.front());
}

void callAccept() {
    std::vector<std::unique_ptr<int>> vec;
    vec.emplace_back(new int(4));
    acceptUniquePtr(vec);
}
```

Reply:
Haha, quite evil, indeed. Lemme try to figure out why the hell I did it this way.

Reply:
Ok, I think the point here was that you don't want to clone RooAbsL's too much, because they contain clones of the pdf and dataset that may have been mangled for optimization. You probably don't want to be doing that all the time, although I must admit that it is a premature optimization, since I haven't timed its impact at all. That is the motivation behind using `unique_ptr`s for the components. The way the classes are built, the `RooSumL` doesn't care about what components it gets, so by definition it cannot create them internally, so they have to be passed in somehow. Forcing the user to call the function with a `std::move` is a way to make them fully realize this.

I could change the type to an rvalue reference to make it clearer from the compiler error that `std::move` is necessary, instead of the obscure error that you get now. Compare the compiler error messages from these two types:

```c++
#include <vector>
#include <memory>
#include <cstdio>

struct Clear {
    Clear(std::vector<std::unique_ptr<int>>&& vec) : vec_(std::move(vec)) {
        printf("number is %d", *vec_[0]);
    }

    std::vector<std::unique_ptr<int>> vec_;
};

struct Obscure {
    Obscure(std::vector<std::unique_ptr<int>> vec) : vec_(std::move(vec)) {
        printf("number is %d", *vec_[0]);
    }

    std::vector<std::unique_ptr<int>> vec_;
};

int main() {
    std::vector<std::unique_ptr<int>> vec;
    vec.emplace_back(new int(4));
    Clear thing(vec);
    Obscure thingy(vec);
}
```

[Godbolt link.](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,fontUsePx:'0',j:2,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:26,positionColumn:2,positionLineNumber:26,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:'%23include+%3Cvector%3E%0A%23include+%3Cmemory%3E%0A%23include+%3Ccstdio%3E%0A%0Astruct+Clear+%7B%0A++++Clear(std::vector%3Cstd::unique_ptr%3Cint%3E%3E%26%26+vec)+:+vec_(std::move(vec))+%7B%0A++++++++printf(%22number+is+%25d%22,+*vec_%5B0%5D)%3B%0A++++%7D%0A%0A++++std::vector%3Cstd::unique_ptr%3Cint%3E%3E+vec_%3B%0A%7D%3B%0A%0Astruct+Obscure+%7B%0A++++Obscure(std::vector%3Cstd::unique_ptr%3Cint%3E%3E+vec)+:+vec_(std::move(vec))+%7B%0A++++++++printf(%22number+is+%25d%22,+*vec_%5B0%5D)%3B%0A++++%7D%0A%0A++++std::vector%3Cstd::unique_ptr%3Cint%3E%3E+vec_%3B%0A%7D%3B%0A%0Aint+main()+%7B%0A++++std::vector%3Cstd::unique_ptr%3Cint%3E%3E+vec%3B%0A++++vec.emplace_back(new+int(4))%3B%0A++++Clear+thing(vec)%3B%0A++++Obscure+thingy(vec)%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%232',t:'0')),header:(),k:50.077875864442966,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:gsnapshot,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:boost,ver:'169'),(name:xsimd,ver:trunk)),options:'-std%3Dc%2B%2B11+-O3+-mavx',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:2),l:'5',n:'0',o:'x86-64+gcc+(trunk)+(Editor+%232,+Compiler+%231)+C%2B%2B',t:'0')),header:(),k:49.92212413555705,l:'4',m:76.82177348551362,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:2,fontScale:10,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+(trunk)+(Compiler+%231)',t:'0')),header:(),l:'4',m:23.17822651448639,n:'0',o:'',s:0,t:'0')),k:49.92212413555705,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Reply:
I added a warning in the documentation, as suggested.